packages/playground/data-liberation/src/xml-api/WP_XML_Tag_Processor.php

```diff
@@ -0,0 +1,2837 @@
+	 * that we're still waiting for more data/
+	 *
+	 * This flag allows us to differentiate between these two
+	 * cases in context-aware APIs such as WP_XML_Processor.
+	 *
+	 * @var bool
+	 */
+	protected $is_incomplete_text_node = false;
+
+	/**
+	 * How many bytes from the original XML document have been read and parsed.
+	 *
+	 * This value points to the latest byte offset in the input document which
+	 * has been already parsed. It is the internal cursor for the Tag Processor
+	 * and updates while scanning through the XML tokens.
+	 *
+	 * @since WP_VERSION
+	 * @var int
+	 */
+	public $bytes_already_parsed = 0;
+
+	/**
+	 * Byte offset in input document where current token starts.
+	 *
+	 * Example:
+	 *
+	 *     <wp:content id="test">...
+	 *     01234
+	 *     - token starts at 0
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @var int|null
+	 */
+	protected $token_starts_at;
+
+	/**
+	 * Byte length of current token.
+	 *
+	 * Example:
+	 *
+	 *     <wp:content id="test">...
+	 *     012345678901234
+	 *     - token length is 14 - 0 = 14
+	 *
+	 *     a <!-- comment --> is a token.
+	 *     0123456789 123456789 123456789
+	 *     - token length is 17 - 2 = 15
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @var int|null
+	 */
+	private $token_length;
+
+	/**
+	 * Byte offset in input document where current tag name starts.
+	 *
+	 * Example:
+	 *
+	 *     <wp:content id="test">...
+	 *     01234
+	 *      - tag name starts at 1
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @var int|null
+	 */
+	private $tag_name_starts_at;
+
+	/**
+	 * Byte length of current tag name.
+	 *
+	 * Example:
+	 *
+	 *     <wp:content id="test">...
+	 *     01234
+	 *      --- tag name length is 3
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @var int|null
+	 */
+	private $tag_name_length;
+
+	/**
+	 * Byte offset into input document where current modifiable text starts.
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @var int
+	 */
+	private $text_starts_at;
+
+	/**
+	 * Byte length of modifiable text.
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @var string
+	 */
+	private $text_length;
+
+	/**
+	 * Whether the current tag is an opening tag, e.g. <wp:content>, or a closing tag, e.g. </wp:content>.
+	 *
+	 * @var bool
+	 */
+	private $is_closing_tag;
+
+	/**
+	 * Stores an explanation for why something failed, if it did.
+	 *
+	 * @see self::get_last_error
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @var string|null
+	 */
+	protected $last_error = null;
+
+	/**
+	 * Lazily-built index of attributes found within an XML tag, keyed by the attribute name.
+	 *
+	 * Example:
+	 *
+	 *     // Supposing the parser is working through this content
+	 *     // and stops after recognizing the `id` attribute.
+	 *     // <wp:content id="test-4" class=outline title="data:text/plain;base64=asdk3nk1j3fo8">
+	 *     //                 ^ parsing will continue from this point.
+	 *     $this->attributes = array(
+	 *         'id' => new WP_HTML_Attribute_Token( 'id', 9, 6, 5, 11, false )
```

Comment:
It's interesting to find `WP_HTML_Attribute_Token` used in the XML tag processor, but I guess the concepts are the same or nearly the same structurally.

Reply:
Yup! But we could also create XML-specific flavors